Katello/katello
# Fixes #29000 - Consume Candlepin events via STOMP

This PR is the result of my POC (#8431) and [RFC](https://community.theforeman.org/t/rfc-goodbye-qpid-hello-artemis-evolving-messaging-in-katello/17016) to implement the Artemis message broker and consume Candlepin messages via the STOMP protocol.

**To test**
Spin up a new (second) development environment using the related installer changes so there's zero-configuration other than checking out this PR once provisioned. Use a box definition like this:

```
centos7-katello-artemis:                                                                                                                                                                        
  box: centos7                                                                                                                                                                                
  ansible:                                                                                                                                                                                    
    playbook: 'playbooks/katello_devel.yml'                                                                                                                                                   
    group: 'devel'                                                                                                                                                                            
    variables:                                                                                                                                                                                
      ssh_forward_agent: true                                                                                                                                                                 
      foreman_installer_module_prs:                                                                                                                                                           
        - theforeman/candlepin/143                                                                                                                                                            
        - theforeman/certs/275                                                                                                                                                                
        - theforeman/katello/327                                                                                                                                                              
        - theforeman/katello_devel/224
```

At some point the installer will fail because the candlepin version in nightly (as of 4/16) is not 3.10. At this point install both RPMs here: http://koji.katello.org/koji/buildinfo?buildID=57101 and run `vagrant provosion` on the box. Installer should complete now. After checking out this PR you'll need to `bundle install` from the foreman dir in order to bring in the stomp gem.

The first request to the server should indicate the the candlepin_event queue has been subscribed to:

```
2020-02-13T19:53:10 [I|app|b3ff7cd3] Katello event daemon started process=31288
2020-02-13T19:53:10 [I|app|b3ff7cd3] Processing by Katello::Api::V2::KatelloPingController#index as */*
2020-02-13T19:53:10 [I|app|b3ff7cd3]   Parameters: {"api_version"=>"v2"}
2020-02-13T19:53:10 [I|app|b3ff7cd3] Subscribed to katello.candlepin.candlepin_events
2020-02-13T19:53:10 [I|kat|b3ff7cd3] Polling Katello Event Queue
```

From here the goal is to make sure that Candlepin events are processed as normal. The filtering block in /etc/candlepin/broker.xml determines which messages will reach our queue to be handled.

- Pool create and delete events are tested by importing and deleting a manifest
- Entitlement create and delete are tested by adding and removing subscriptions from a content host
- Compliance events are tested by adding/removing subscriptions or changing system purpose on a content host

There's debug logging for the received events:

```
2020-02-13T20:07:44 [D|kat|b3ff7cd3] message received from subscriptions queue 
2020-02-13T20:07:44 [D|kat|b3ff7cd3] message subject: compliance.created
```

Finally, the listener must be resilient. Get creative and try to break it (randomly restarting tomcat for example), and study the code for weaknesses. If tomcat is down the rails server should continue to serve requests - ie it will not hang indefinitely waiting for the event listener to start. But once tomcat comes back up, you should see that it once again subscribes and handles messages.

**Design**
My main goal was to maintain the behaviors we had with the qpid implementation in terms of resiliency and the messages that reach the listener. I've also created a pattern so that the listener's connection is set up with configuration settings in engine.rb This makes the listener completely unaware of the implementation details - it just subscribes and handles messages. 

The code also makes it easy to add new types of messaging connections so this code will not have to be rewritten from the ground up if we change message brokers in the future. New connection types can be added and instantiated using the `Katello::Messaging::Connection` class. If we can't extract qpid's arrow from our hearts, we can add a new 'QpidConnection' for example.
